File: packages/py/core/src/uri_resolution/resolvers/extendable/uri_resolver_wrapper.py
from __future__ import annotations
from typing import Awaitable, Union

from ....interfaces import UriResolverInterface
from ....manifest import deserialize_web3_api_manifest
from ....types import Uri, Client, InvokeHandler
from ...core import UriResolver, UriResolutionStack, UriResolutionResult
from ....uri_resolution.resolvers.get_env_from_uri_or_resolution import get_env_from_uri_or_resolution_stack


class UriResolverWrapper(UriResolver):
    def __init__(
        self, implementation_uri: Uri, create_api: CreateApiFunc, deserialize_options: DeserializeManifestOptions = None
    ):
        self.implementation_uri = implementation_uri
        self.create_api = create_api
        self.deserialize_options = deserialize_options

    @property
    def name(self) -> str:
        return type(self).__name__

    async def resolve_uri(
        self, uri: Uri, client: Client, cache: ApiCache, resolution_path: UriResolutionStack
    ) -> Awaitable(UriResolutionResult):
        result = await try_resolve_uri_with_implementation(uri, self.implementation_uri, client.invoke)

        if not result:
            return UriResolutionResult(uri)

        if result.get("uri"):
            return UriResolutionResult(uri=Uri(result.get("uri")))
        elif result.get("manifest"):
            # We've found our manifest at the current implementation,
            # meaning the URI resolver can also be used as an API resolver
            manifest = deserialize_web3_api_manifest(result.get("manifest"), self.deserialize_options)

            environment = get_env_from_uri_or_resolution_stack(uri, resolution_path, client)

            api = self.create_api(uri, manifest, self.implementation_uri.uri, environment)

            return UriResolutionResult(uri, api)

        return UriResolutionResult(uri)


async def try_resolve_uri_with_implementation(
    uri: Uri, implementation_uri: Uri, invoke: InvokeHandler["invoke"]
) -> Awaitable(Union[UriResolverInterface.MaybeUriOrManifest, None]):
    result = await UriResolverInterface.try_resolve_uri(invoke, implementation_uri, uri)
    data = result.data

    # If nothing was returned, the URI is not supported
    if not data or (not data.get("uri") and not data.get("manifest")):
        return None

    return data

File: packages/js/core/src/uri-resolution/resolvers/UriResolverWrapper.ts
import { UriResolverInterface } from "../../interfaces";
import { Uri, WrapperCache, Client, Result, Ok, Err, Wrapper } from "../..";
import {
  IUriResolver,
  IUriResolutionStep,
  IUriResolutionResponse,
  UriResolutionResponse,
} from "../core";
import { CreateWrapperFunc } from "./extendable/types/CreateWrapperFunc";
import { getEnvFromUriOrResolutionPath } from "./getEnvFromUriOrResolutionPath";

import {
  DeserializeManifestOptions,
  deserializeWrapManifest,
} from "@polywrap/wrap-manifest-types-js";
import { Tracer } from "@polywrap/tracing-js";
import { initWrapper, IWrapPackage } from "../../types";

export class UriResolverWrapper implements IUriResolver<unknown> {
  constructor(
    public readonly implementationUri: Uri,
    private readonly createWrapper: CreateWrapperFunc,
    private readonly deserializeOptions?: DeserializeManifestOptions
  ) {}

  public get name(): string {
    return `${UriResolverWrapper.name}: (${this.implementationUri.uri})`;
  }

  async tryResolveToWrapper(
    uri: Uri,
    client: Client,
    cache: WrapperCache,
    resolutionPath: IUriResolutionStep<unknown>[]
  ): Promise<IUriResolutionResponse<unknown>> {
    const { result, history } = await tryResolveUriWithImplementation(
      uri,
      this.implementationUri,
      client
    );

    if (!result.ok) {
      return UriResolutionResponse.err(result.error, history);
    }

    const uriOrManifest = result.value;

    if (uriOrManifest?.uri) {
      return UriResolutionResponse.ok(new Uri(uriOrManifest.uri), history);
    } else if (uriOrManifest?.manifest) {
      // We've found our manifest at the current implementation,
      // meaning the URI resolver can also be used as an Wrapper resolver
      const manifest = await deserializeWrapManifest(
        uriOrManifest.manifest,
        this.deserializeOptions
      );

      const environment = getEnvFromUriOrResolutionPath(
        uri,
        resolutionPath,
        client
      );
      const wrapper = this.createWrapper(
        uri,
        manifest,
        this.implementationUri.uri,
        environment
      );

      return UriResolutionResponse.ok(wrapper, history);
    }

    return UriResolutionResponse.ok(uri, history);
  }
}

const tryResolveUriWithImplementation = async (
  uri: Uri,
  implementationUri: Uri,
  client: Client
): Promise<{
  result: Result<UriResolverInterface.MaybeUriOrManifest | undefined, unknown>;
  history?: IUriResolutionStep<unknown>[];
}> => {
  const response = await client.tryResolveToWrapper(implementationUri);

  if (!response.result.ok) {
    return {
      result: Err(response.result.error),
      history: response.history,
    };
  }

  const uriPackageOrWrapper = response.result.value;

  if (uriPackageOrWrapper.type === "uri") {
    const lastFoundUri = uriPackageOrWrapper.uri as Uri;

    return {
      result: Err(
        `While resolving ${uri.uri} with URI resolver extension ${implementationUri.uri}, the extension could not be fully resolved. Last found URI is ${lastFoundUri.uri}`
      ),
      history: response.history,
    };
  }

  let wrapperOrPackage: IWrapPackage | Wrapper;

  if (uriPackageOrWrapper.type === "package") {
    wrapperOrPackage = uriPackageOrWrapper.package;
  } else {
    wrapperOrPackage = uriPackageOrWrapper.wrapper;
  }

  const wrapper = await initWrapper(wrapperOrPackage, client);

  const invokeResult = await client.invokeWrapper<UriResolverInterface.MaybeUriOrManifest>(
    {
      wrapper,
      uri: implementationUri.uri,
      method: "tryResolveUri",
      args: {
        authority: uri.authority,
        path: uri.path,
      },
    }
  );

  const { data, error } = invokeResult;

  const uriOrManifest = data as UriResolverInterface.MaybeUriOrManifest;

  if (error) {
    return {
      result: Err(error),
      history: [],
    };
  }
  Tracer.addEvent("continue", implementationUri.uri);

  return {
    result: Ok(uriOrManifest ?? undefined),
    history: [],
  };
};
